src/sqlancer/ASTBasedReducer.java

```diff
@@ -0,0 +1,470 @@
+            Expression elseExpr = expr.getElseExpression();
+
+            expressionReduce(expr, switchExpr, (parent, sw) -> {
+                parent.setSwitchExpression(sw);
+                updateStatements();
+            });
+            expressionReduce(expr, elseExpr, (parent, els) -> {
+                parent.setElseExpression(els);
+                updateStatements();
+            });
+
+            super.visit(expr);
+        }
+
+        @Override
+        public void visit(WhenClause whenClause) {
+            Expression when = whenClause.getWhenExpression();
+            Expression then = whenClause.getThenExpression();
+
+            expressionReduce(whenClause, when, (wc, w) -> {
+                wc.setWhenExpression(w);
+                updateStatements();
+            });
+
+            expressionReduce(whenClause, then, (wc, t) -> {
+                wc.setThenExpression(t);
+                updateStatements();
+            });
+
+            super.visit(whenClause);
+        }
+
+        @Override
+        public void visit(Parenthesis parenthesis) {
+            Expression closedExpr = parenthesis.getExpression();
+            expressionReduce(parenthesis, closedExpr, (p, s) -> {
+                p.setExpression(s);
+                updateStatements();
+            });
+            closedExpr.accept(this);
+        }
+
+        @Override
+        public void visit(Function function) {
+            super.visit(function);
+        }
+
+        @Override
+        public void visit(ExpressionList expressionList) {
+            List<Expression> expressions = expressionList.getExpressions();
+            listElementRemovingReduce(expressionList, expressions, (l, es) -> {
+                l.setExpressions(es);
+                updateStatements();
+            });
+            expressions = expressionList.getExpressions();
+            for (int i = 0; i < expressions.size(); i++) {
+                Expression expr = expressions.get(i);
+                int index = i;
+                expressionReduce(expressions, expr, (l, e) -> {
+                    l.set(index, e);
+                    updateStatements();
+                });
+            }
+            super.visit(expressionList);
+        }
+
+        @Override
+        public void visit(SelectExpressionItem selectExpressionItem) {
+            Expression expr = selectExpressionItem.getExpression();
+            expressionReduce(selectExpressionItem, expr, (item, e) -> {
+                item.setExpression(e);
+                updateStatements();
+            });
+            super.visit(selectExpressionItem);
+        }
+    };
+
+    SelectVisitorAdapter selectReducerVisitor = new SelectVisitorAdapter() {
+        private final String[] removeList = { "Limit", "Offset", "Where", "Having", "GroupBy", "Distinct",
+                "OrderByElements", "Joins" };
+        private final String[] pullUpList = { "Where", "Having", "FromItem", "SelectItems", "GroupBy", "Joins" };
```

Comment:
can we support common table expression (https://dev.mysql.com/doc/refman/8.0/en/with.html) here
and derived tables (https://dev.mysql.com/doc/refman/8.0/en/derived-tables.html)

Reply:
Theoretically both could be handled. 
For derived tables I've added support for it. :
Given a statement:
```sql 
SELECT AVG(sum_column1)
FROM (SELECT SUM(column1) AS sum_column1 FROM t1 GROUP BY column1 LIMIT 32 OFFSET 128) AS t1;
```
The result is:
``` sql 
SELECT AVG(sum_column1) FROM (SELECT SUM(column1) AS sum_column1 FROM t1) AS t1;
```
As for common table expression, it hasn't be supported yet.